dex/src/main/java/io/crate/data/CollectingBatchIterator.java

```diff
@@ -42,12 +45,23 @@
 public class CollectingBatchIterator<A> implements BatchIterator {
 
     private final BatchIterator source;
-    private final Collector<Row, A, ? extends Iterable<Row>> collector;
+    private final Function<BatchIterator, CompletableFuture<? extends Iterable<Row>>> consumer;
     private final RowColumns rowData;
 
     private Iterator<Row> it = Collections.emptyIterator();
     private CompletableFuture<? extends Iterable<Row>> resultFuture;
 
+    private CollectingBatchIterator(BatchIterator source, Function<BatchIterator, CompletableFuture<? extends Iterable<Row>>> consumer,
+                                    int numCols) {
+        this.source = source;
+        this.consumer = consumer;
+        this.rowData = new RowColumns(numCols);
+    }
+
+    private CollectingBatchIterator(BatchIterator source, Collector<Row, A, ? extends Iterable<Row>> collector, int numCols) {
```

Comment:
Could the public `newInstance` factory method already convert the collector into a function? Would remove the need for this extra constructor here